vulcan/lib/orchestration.rb

```diff
@@ -18,8 +20,13 @@ def run_until_done!(storage)
       [].tap do |ran|
         i = 0
         until (runnables = next_runnable_build_targets(storage)).empty? || (i += 1) > MAX_RUNNABLE
-          run!(storage: storage, build_target: runnables.first)
-          ran << runnables.first
+          begin
+            run!(storage: storage, build_target: runnables.first)
```

Comment:
Hmm... I gotta say that I don't like this approach at all.  This approach would have the runner keep executing in the face of errors, which likely will just result in it trying the same cell over and over until i > MAX_RUNNABLE.  Furthermore, I feel like catching errors here is the wrong place to handle.  How does a caller to `run_until_done!` know if something went wrong?  Placing it in an errors hash seems wrong as well.  Who clears the errors hash?  How do you know if this run was safe but the previous wasn't?  I don't like this approach at all I think.

Reply:
If the idea is that you want to attempt to run _other_ nodes before raising an exception, I'd recommend iterating the runnables value (it is an array itself), trying to run each of those, collecting errors, and then immediately throwing possibly a new error type that collects inner errors by step.

Reply:
Thanks for the suggestions! Okay, I'll rework this. The idea was that users would be able to see exactly which step(s) failed and why, but with the previous behavior, that was impossible -- the 500 didn't report back the step name or any related context, just a stacktrace without an identifiable relation to a step (i.e. `script.py` doesn't correlate to a specific step). It also seemed undesirable from a debugging standpoint if multiple steps would have failed (i.e. parallel steps), but having to iterate through them one-at-a-time.

Also, I thought the overall philosophy we were following was for the UI to be thin / simple, so thus the `status` response would encapsulate all the information about each step? But if it's preferable to have the controller and orchestrator throw exceptions / 500's instead of updating the status elements, and leave the UI to figure out how that information ties together with the "good" status, I'll just re-think how to present it.

Reply:
absolutely.  I agree with the direction of this pr, only the implementation am I concerned about it.  I think we can get the same exact effect, but keep the state local, and return it back to the controller caller, who can unwrap it.

Reply:
@corps , could you take another look and see if the updated approach works better?